src/catalog/external.rs

```diff
@@ -0,0 +1,198 @@
+use crate::catalog::{
+    not_impl, CatalogResult, CatalogStore, FunctionStore, SchemaStore, TableStore,
+};
+use crate::repository::interface::{
+    AllDatabaseFunctionsResult, CollectionRecord, DatabaseRecord,
+    DroppedTableDeletionStatus, DroppedTablesResult, TableId, TableRecord,
+    TableVersionId, TableVersionsResult,
+};
+use crate::wasm_udf::data_types::CreateFunctionDetails;
+use arrow_schema::Schema;
+use clade::catalog::CatalogReference;
+use clade::schema::schema_store_service_client::SchemaStoreServiceClient;
+use clade::schema::ListSchemaResponse;
+use tonic::transport::{channel::Channel, Error};
+use tonic::Request;
+use uuid::Uuid;
+
+// An external store, facilitated via a remote clade server implementation
+#[derive(Clone)]
+pub struct ExternalStore {
+    client: SchemaStoreServiceClient<Channel>,
+}
+
+impl ExternalStore {
+    // Create a new external store implementing the clade interface
+    pub async fn new(dsn: String) -> Result<Self, Error> {
+        println!("Client dsn is {dsn}");
+        let client = SchemaStoreServiceClient::connect(dsn).await?;
+        Ok(Self { client })
+    }
+
+    // Tonic client implementations always end up needing mut references, and apparently the way
+    // to go is cloning a client instance instead of introducing synchronization primitives:
+    // https://github.com/hyperium/tonic/issues/33#issuecomment-538154015
+    fn client(&self) -> SchemaStoreServiceClient<Channel> {
+        self.client.clone()
+    }
+}
+
+#[tonic::async_trait]
+impl CatalogStore for ExternalStore {
+    async fn create(&self, _name: &str) -> CatalogResult<()> {
+        not_impl()
+    }
+
+    async fn get(&self, _name: &str) -> CatalogResult<DatabaseRecord> {
+        not_impl()
+    }
+
+    async fn delete(&self, _name: &str) -> CatalogResult<()> {
+        not_impl()
+    }
+}
+
+#[tonic::async_trait]
+impl SchemaStore for ExternalStore {
+    async fn create(&self, _catalog_name: &str, _schema_name: &str) -> CatalogResult<()> {
+        not_impl()
+    }
+
+    async fn list(&self, catalog_name: &str) -> CatalogResult<ListSchemaResponse> {
+        let req = Request::new(CatalogReference {
+            catalog_name: catalog_name.to_string(),
+        });
+
+        let response = self.client().list(req).await?;
```

Comment:
That is correct: all the autogen client methods take `&mut self` and it seems better to avoid introducing a Mutex or other locking mechanisms here and just clone the client on every call, since that is apparently cheap.